src/wallet/migrate.cpp

```diff
@@ -1,13 +1,601 @@
+            {
+                DataRecord record(rec_hdr);
+                s >> record;
+                records.emplace_back(record);
+                to_jump += rec_hdr.len;
+                break;
+            }
+            case RecordType::DUPLICATE:
+            case RecordType::OVERFLOW_DATA:
+            {
+                OverflowRecord record(rec_hdr);
+                s >> record;
+                records.emplace_back(record);
+                to_jump += OverflowRecord::SIZE;
+                break;
+            }
+            default:
+                throw std::runtime_error("Unknown record type in records page");
+            }
+
+            // Go back to the indexes
+            s.seek(-to_jump, SEEK_CUR);
+        }
+    }
+};
+
+/** A page containing overflow data */
+class OverflowPage
+{
+public:
+    OverflowPage(const PageHeader& header) : m_header(header) {}
+    OverflowPage() = delete;
+
+    PageHeader m_header;
+
+    // BDB overloads some page fields to store overflow page data
+    // hf_offset contains the length of the overflow data stored on this page
+    // entries contains a reference count for references to this item
+
+    // The overflow data itself. Begins immediately following header
+    std::vector<std::byte> data;
+
+    template<typename Stream>
+    void Unserialize(Stream& s)
+    {
+        data.resize(m_header.hf_offset);
+        s.read(AsWritableBytes(Span(data.data(), data.size())));
+    }
+};
+
+/** A page of records in the database */
+class InternalPage
+{
+public:
+    InternalPage(const PageHeader& header) : m_header(header) {}
+    InternalPage() = delete;
+
+    PageHeader m_header;
+
+    std::vector<uint16_t> indexes;
+    std::vector<InternalRecord> records;
+
+    template<typename Stream>
+    void Unserialize(Stream& s)
+    {
+        // Current position within the page
+        int64_t pos = PageHeader::SIZE;
+
+        // Get the items
+        for (uint32_t i = 0; i < m_header.entries; ++i) {
+            // Get the index
+            uint16_t index;
+            s >> index;
+            indexes.push_back(index);
+            pos += sizeof(uint16_t);
+
+            // Go to the offset from the index
+            int64_t to_jump = index - pos;
+            s.ignore(to_jump);
+
+            // Read the record
+            RecordHeader rec_hdr;
+            s >> rec_hdr;
+            to_jump += RecordHeader::SIZE;
+
+            if (rec_hdr.GetRealType() != RecordType::KEYDATA) {
+                throw std::runtime_error("Unknown record type in internal page");
+            }
+            InternalRecord record(rec_hdr);
+            s >> record;
+            records.emplace_back(record);
+            to_jump += InternalRecord::FIXED_SIZE + rec_hdr.len;
+
+            // Go back to the indexes
+            s.seek(-to_jump, SEEK_CUR);
+
+        }
+    }
+};
+
+static void SeekToPage(AutoFile& s, uint32_t page_num, uint32_t page_size)
+{
+    size_t pos = page_num * page_size;
```

Comment:
calculation of two 32 bit integers, you need to use a cast

Reply:
aside: might be better to not use size_t at all for file offsets, as the type is 32 bit on usual 32 bit platforms